### PR TITLE
Add auth

### PR DESCRIPTION
**Pull request by djobbo**

### CODE REVIEW

File: app/app/login/github/callback/route.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      headers: {
+        Location: "/",
+      },
+    });
+  } catch (e) {
+    // the specific error message depends on the provider
+    if (e instanceof OAuth2RequestError) {
+      // invalid code
+      return new Response(null, {
+        status: 400,
```
Comment (JoachimFarabet) [inline app/app/login/github/callback/route.ts:88]: You should make custom status, no magic number

File: app/app/login/github/callback/route.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      });
+    }
+
+    return new Response(null, {
+      status: 500,
```
Comment (JoachimFarabet) [inline app/app/login/github/callback/route.ts:93]: Same as above

File: app/app/login/github/route.ts [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  const state = generateState();
+  const url = await github.createAuthorizationURL(state, {
+    scopes: ["user:email"],
+  });
+
+  cookies().set("github_oauth_state", state, {
+    path: "/",
+    secure: process.env.NODE_ENV === "production",
+    httpOnly: true,
+    maxAge: 60 * 10,
```
Comment (JoachimFarabet) [inline app/app/login/github/route.ts:15]: no magic number

File: app/app/login/page.tsx [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export default async function Page() {
+  return (
+    <>
+      <h1>Sign in</h1>
+      <a href="/login/github">Sign in with GitHub</a>
+    </>
```
Comment (JoachimFarabet) [inline app/app/login/page.tsx:6]: every page should start with a `<section>`

File: app/app/page.tsx [hunk 5]
```diff
@@ -1,3 +1,10 @@
-export default function Home() {
-  return <main></main>;
+import { getCurrentSession } from "@/auth";
+import Link from "next/link";
+
+export default async function Home() {
+  const { user } = await getCurrentSession();
+
+  return (
+    <>{user ? <>{user.username}</> : <Link href="/login">Sign in</Link>}</>
```
Comment (JoachimFarabet) [inline app/app/page.tsx:8]: You can remove those fragments

File: app/auth/github.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+import { GitHub } from "arctic";
+
+const GITHUB_CLIENT_ID = process.env.GITHUB_CLIENT_ID!;
+const GITHUB_CLIENT_SECRET = process.env.GITHUB_CLIENT_SECRET!;
+
+export const github = new GitHub(GITHUB_CLIENT_ID!, GITHUB_CLIENT_SECRET!);
```
Comment (djobbo) [inline app/auth/github.ts:6]: no `!` 😠

File: app/auth/lucia.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  },
+  getUserAttributes: (attributes) => {
+    return {
+      githubId: attributes.github_id,
+      username: attributes.username,
+    };
+  },
+});
+
+// IMPORTANT!
```
Comment (djobbo) [inline app/auth/lucia.ts:31]: is `IMPORTANT!` that important?

**Review by JoachimFarabet — APPROVED** [review]
LGTM

**Pull request merged**